sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1660 @@
+`handler::depends_on()` will also create a graph edge when passed an event
+returned from a queue submission captured by a queue recording to the same graph.
+|===
+
+==== Queue Recording API
+
+When using the record & replay API to construct a graph by recording a queue,
+nodes and edges are captured as follows.
+
+Table {counter: tableNumber}. Recorded Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| A node in a queue recorded graph represents a command group submission to the
+device associated with the queue begin recorded. Each submission encompasses
+either one or both of a.) some data movement, b.) a single asynchronous kernel
+launch. Nodes cannot define forward edges, only backwards. That is, kernels can
+only create dependencies on command-groups that have already been submitted.
+This means that transparently a node can depend on a previously recorded graph
+(sub-graph), which works by creating edges to the individual nodes in the old
+graph. Explicit memory operations without kernels, such as a memory copy, are
+still classed as nodes under this definition, as the
+{explicit-memory-ops}[SYCL 2020 specification states] that these can be seen as
+specialized kernels executing on the device.
+
+| Edge
+| An edge in a queue recorded graph is expressed through command group
+dependencies in one of two ways. Firstly, through buffer accessors that
+represent data dependencies between two command groups captured as nodes.
+Secondly, by using the `handler::depends_on()` mechanism inside a command group
+captured as a node. However, for an event passed to `handler::depends_on()` to
+create an edge, it must be an event returned from a queue
+submission captured by the same graph. Otherwise, a synchronous error will be
+thrown with error code `invalid`. `handler::depends_on()` can be
+used to express edges when a user is working with USM memory rather than SYCL
+buffers.
+|===
+
+==== Sub-Graph
+
+A node in a graph can take the form of a nested sub-graph. This occurs when
+a command-group submission that invokes `handler::ext_oneapi_graph()` with an
+executable graph object is added to the graph as a node.
+
+=== API Modifications
+
+[source, c++]
+----
+namespace sycl {
+namespace ext::oneapi::experimental {
+
+// State of a queue, returned by queue::ext_oneapi_get_state()
+enum class queue_state {
+  executing,
+  recording
+};
+
+namespace property {
+
+namespace graph {
+
+class no_cycle_check {
+  public:
+    no_cycle_check() = default;
+};
+
+class no_host_copy {
+public:
+  no_host_copy() = default;
+};
+} // namespace graph
+
+namespace node {
+
+class depends_on {
+  public:
+    template<typename... NodeTN>
+    depends_on(NodeTN... nodes);
+};
+
+} // namespace node
+} // namespace property
+
+// Device query for level of support
+namespace info {
+namespace device {
+struct graphs_support;
+
+enum class graph_support_level {
+  unsupported,
+  native,
+  emulated
+};
```

Comment:
I think it would be better to have `graph_support_level` in the `info` namespace rather than `info::device`. Partially because that's what other info return types do and partially to ensure people don't mistake it for an information query type, like `graphs_support`.

Reply:
Good spot, updated in new commit https://github.com/intel/llvm/pull/5626/commits/68c5b42999ccb861849d593ef15183885569f6b8